### ISSUE TITLE
Deep Double Descent: Where Bigger Models and More Data Hurt

### ISSUE THREAD

**Issue by yonesuke0716** [TeamAidemy/ds-paper-summaries#43]
Nakkiran, Preetum, Gal Kaplun, Yamini Bansal, Tristan Yang, Boaz Barak, and Ilya Sutskever. 2019. “Deep Double Descent: Where Bigger Models and More Data Hurt.” ArXiv:1912.02292 [Cs, Stat], December. https://arxiv.org/abs/1912.02292.

- 機械学習において複雑なモデルはBias-Variance Trade-offによって過学習に陥りやすいとされてきた。しかし、Deep Learningにおいては過学習を超えてLossが再度降下する現象「Double Descent」があることを発見した。
- モデルのパラメータ数、学習回数、サンプルサイズ、ラベルノイズの条件を様々に組み合わせた実験を行うことで、Double-Descentの現象を説明する仮説を立てた。
- モデルの複雑度が十分高いほど過学習を超えて汎化を獲得しやすく、学習回数も十分多いほど汎化を獲得していく。

## Abstract
>We show that a variety of modern deep learning tasks exhibit a “double-descent”phenomenon where, as we increase model size, performance first gets worse and then gets better. Moreover, we show that double descent occurs not just as a function of model size, but also as a function of the number of training epochs. We unify the above phenomena by defining a new complexity measure we call the effective model complexity and conjecture a generalized double descent with respect to this measure. Furthermore, our notion of model complexity allows us to identify certain regimes where increasing (even quadrupling) the number of train samples actually hurts test performance.

(DeepL翻訳)
我々は、現代の様々な深層学習課題において、モデルサイズを大きくすると、まず性能が低下し、次に良くなるという「二重降下」現象が起こることを示す。さらに、二重降下はモデルサイズの関数としてだけでなく、学習エポック数の関数としても発生することを示す。我々は、有効モデル複雑度と呼ぶ新しい複雑度指標を定義することで上記の現象を統一し、この指標に関して一般化された二重降下が起こることを予言する。さらに、このモデルの複雑さの概念により、訓練サンプル数を増やすと（4倍でも）テスト性能が低下する特定の領域を特定することができる。

## コード
https://gitlab.com/harvard-machine-learning/double-descent/tree/master

## 解決した課題/先行研究との比較
これまで機械学習における汎化と過学習については、Bias-Variance Trade-offという概念で説明されてきた。
<img width="355" alt="スクリーンショット 2023-05-17 11 15 21" src="https://github.com/TeamAidemy/ds-paper-summaries-internal/assets/92288672/86d8217b-1496-42d7-98ee-e754a8cfebe1">

- Hastie et al., 2009 "The Elements of Statistical Learning"よりFigure 2.11.を転載

Biasは誤差の偏りを表し、Varianceは誤差の分散（ばらつき）を表す。

モデルの複雑さが高いほどBiasは小さくなるがVarianceは大きくなり、過学習が進む。

一方でモデルの複雑さが少ないほどBiasは大きくなりVarianceが小さくなるため、学習不足となる。

しかし機械学習の研究では「複雑なモデルでも過学習を超えて再び汎化する」ケースが存在する。

その事例としてとDouble Descentとという現象が発見されている。

Double Descentとは、最初test errorが単調に減少した後(アンダーパラメータ)、test errorが増加(過学習)、そして再びtest errorが下がり出す二重降下現象のことである。

例えば機械学習の分野では、Double Descentの一般化を仮説する論文として、以下が紹介されている。

[Reconciling modern machine learning and the bias-variance trade-off](https://arxiv.org/abs/1812.11118)

本論文ではDeep LearningにおいてもDouble Descentが一般的に起こりうるものであり、Deep Double Descentとして一般的に説明できると仮説を立てている。

<img width="540" alt="スクリーンショット 2023-06-12 20 29 43" src="https://github.com/TeamAidemy/ds-paper-summaries-internal/assets/92288672/49c6c7d3-79e5-4f0b-abcc-09930fcca9c1">

特にどのような条件でDeep Double Descentが発生しやすいのか、様々な条件下での実験を行い検証している。

## 実験方法
学習が完了するまでのtest errorを計測する。

計測中にDouble Descentがどのようにして起こるか、様々なデータセット、アーキテクチャ、最適化アルゴリズムのいくつかの選択を考慮して、モデルのパラメータ数、エポック数、トレーニングサンプル数を変化させて検証している。

- Model-wise Double Descent
　→エポック数を固定し、モデルのパラメータ数毎に比較。
- Epoch-wise Double Descent
　→モデルのパラメータ数を固定し、エポック数毎に比較。
- Sample-wise Non-monotonicity
　→モデルのパラメータ数を固定。サンプルサイズ毎に比較。

またラベルノイズ（ラベルの付け間違い）を混ぜた条件でも実施している。
現実世界でもラベルの付け間違いは起こりうるものであり、それを想定した検証である。

## 結果
上記実験の結果を、以下のテーブルにまとめている。

<img width="695" alt="スクリーンショット 2023-05-12 20 43 43" src="https://github.com/TeamAidemy/ds-paper-summaries-internal/assets/92288672/24b7e606-ef3a-4dbe-ab6b-9186e431ad4f">

Double Descentの項にチェックがついているものがDouble-Descentの現象を確認できたパターンとなっている。

### Model-wise Double Descent
<img width="713" alt="スクリーンショット 2023-05-26 17 23 04" src="https://github.com/TeamAidemy/ds-paper-summaries-internal/assets/92288672/d2880c55-8698-4a77-a8f9-9e1cf7cbe213">

ラベルノイズが0%ではDouble Descentが確認できないが、10%, 20%と増加するにつれてDouble Descentが強調されて確認しやすくなる。

<img width="720" alt="スクリーンショット 2023-05-26 17 25 40" src="https://github.com/TeamAidemy/ds-paper-summaries-internal/assets/92288672/fc3f24b6-11c4-48fd-a75d-4d2f7a6ff86a">

Augmentationは、ラベルノイズの付与率に関係なく、Double Descentのピークを右にシフトさせる。

### Epoch-wise Double Descent

<img width="697" alt="スクリーンショット 2023-05-26 17 29 33" src="https://github.com/TeamAidemy/ds-paper-summaries-internal/assets/92288672/61c08b7b-d653-4bab-a995-1c4550015545">

大型なモデルの場合はDouble Descentが発生している。

中型なモデルの場合は早期に止めた方が良い古典的なU字型のtest errorを描いている。

小型なモデルの場合は単調に減少する「アンダーパラメータ」なtest errorを描いている。

<img width="689" alt="スクリーンショット 2023-05-26 17 30 19" src="https://github.com/TeamAidemy/ds-paper-summaries-internal/assets/92288672/350a7d0b-ea2a-436f-baf6-f4366061f850">

ラベルノイズを付与することで、Double Descentの現象が強調される。

これはModel-wise の結果と一致する。

### Sample-wise Non-monotonicity
<img width="365" alt="スクリーンショット 2023-05-26 21 48 31" src="https://github.com/TeamAidemy/ds-paper-summaries-internal/assets/92288672/b0ec285b-7f41-414e-a55c-ff5ec580118a">

データのサンプルサイズの大小関係なく、Double Descentは確認できる。

Model-wiseのAugmentationの結果と同じく、サンプルサイズが多いほどDouble Descentのピークは右にシフトする。

<img width="307" alt="スクリーンショット 2023-05-26 21 45 40" src="https://github.com/TeamAidemy/ds-paper-summaries-internal/assets/92288672/8b95c3fd-cd80-4a16-9f12-d62099ab111f">

データのサンプルサイズは多ければ多いほどtest errorは下がる。

## 残された課題・議論・感想

※ 感想
- 汎化な学習モデルを作成する際に、パラメータ数やエポック数、データのサンプルサイズの用意の仕方で参考になった。
- これまではエポック数についてはearlystoppingを採用することが多かったが、学習環境のリソースが許す限りエポック数を多めに実行してみると新たな発見があるかもしれないと思った。

## 重要な引用

- Mikhail Belkin, Daniel Hsu, and Ji Xu. Two models of double descent for weak features. arXiv preprint arXiv:1903.07571, 2019.

- Koby Bibas, Yaniv Fogel, and Meir Feder. A new look at an old problem: A universal learning approach to linear regression. arXiv preprint arXiv:1905.04708, 2019.

- Mauro Cettolo, Christian Girardi, and Marcello Federico. Wit3: Web inventory of transcribed and translated talks. In Proceedings of the 16th Conference of the European Association for Machine Translation (EAMT), pp. 261–268, Trento, Italy, May 2012.

- Mario Geiger, Arthur Jacot, Stefano Spigler, Franck Gabriel, Levent Sagun, Stephane d’Ascoli, ´Giulio Biroli, Clement Hongler, and Matthieu Wyart. Scaling description of generalization with ´number of parameters in deep learning. arXiv preprint arXiv:1901.01608, 2019a.

- Mario Geiger, Stefano Spigler, Stephane d’Ascoli, Levent Sagun, Marco Baity-Jesi, Giulio Biroli, ´and Matthieu Wyart. Jamming transition as a paradigm to understand the loss landscape of deep neural networks. Physical Review E, 100(1):012115, 2019b.

- Ian J Goodfellow, Jonathon Shlens, and Christian Szegedy. Explaining and harnessing adversarial examples. arXiv preprint arXiv:1412.6572, 2014.

- Trevor Hastie, Robert Tibshirani, Jerome Friedman, and James Franklin. The elements of statistical learning: data mining, inference and prediction. The Mathematical Intelligencer, 27(2):83–85,2005.

- Trevor Hastie, Andrea Montanari, Saharon Rosset, and Ryan J Tibshirani. Surprises in highdimensional ridgeless least squares interpolation. arXiv preprint arXiv:1903.08560, 2019.

- Kaiming He, Xiangyu Zhang, Shaoqing Ren, and Jian Sun. Identity mappings in deep residual networks. In European conference on computer vision, pp. 630–645. Springer, 2016.

- Yanping Huang, Yonglong Cheng, Dehao Chen, HyoukJoong Lee, Jiquan Ngiam, Quoc V. Le, and Zhifeng Chen. Gpipe: Efficient training of giant neural networks using pipeline parallelism. CoRR, abs/1811.06965, 2018. URL http://arxiv.org/abs/1811.06965.

- Alex Krizhevsky. Learning multiple layers of features from tiny images. Technical report, 2009.

- Alex Krizhevsky, Ilya Sutskever, and Geoffrey E Hinton. Imagenet classification with deep convolutional neural networks. In Advances in neural information processing systems, pp. 1097–1105,2012.

- Aleksander Madry, Aleksandar Makelov, Ludwig Schmidt, Dimitris Tsipras, and Adrian Vladu. Towards deep learning models resistant to adversarial attacks. arXiv preprint arXiv:1706.06083,2017.

- Song Mei and Andrea Montanari. The generalization error of random features regression: Precise asymptotics and double descent curve. arXiv preprint arXiv:1908.05355, 2019.

- Partha P. Mitra. Understanding overfitting peaks in generalization error: Analytical risk curves for l2 and l1 penalized interpolation. ArXiv, abs/1906.03667, 2019.

- Vidya Muthukumar, Kailas Vodrahalli, and Anant Sahai. Harmless interpolation of noisy data in regression. arXiv preprint arXiv:1903.09139, 2019.

- Preetum Nakkiran, Gal Kaplun, Dimitris Kalimeris, Tristan Yang, Benjamin L Edelman, Fred Zhang, and Boaz Barak. Sgd on neural networks learns functions of increasing complexity. arXiv preprint arXiv:1905.11604, 2019.

- Manfred Opper. Statistical mechanics of learning: Generalization. The Handbook of Brain Theory and Neural Networks, 922-925., 1995.

- Manfred Opper. Learning to generalize. Frontiers of Life, 3(part 2), pp.763-775., 2001.

## 参考情報
Attentionモデルにおける初期オーバーフィット後の汎化性能の獲得現象「Grokking」についての論文の要約は[こちら](https://github.com/TeamAidemy/ds-paper-summaries-internal/issues/31#issue-1641531746)
